### PR TITLE
feat(cache): add default param to cache.get()

### PR DESCRIPTION
**Pull request by dbowling**
- `cache.get()` gains second parameter, enabling a default value to be defined on first call for a given key

Closes issue: #3276

### CODE REVIEW

**Comment by dbowling** [comment]
@straker I had to spread the params of `get()` to account for the possibility of an `undefined` default value. This does change the way the cache behaves a bit though--any params passed in excess of the first were ignored, and passing no params gave you some unexpected behavior.

It could be smart to add a guard to `get()` and throw if there are the wrong number of args, considering how cache methods can be a real pain if they aren't predictable. Adding a throw, though, feels like it is at a greater risk of introducing a breaking change than adding this feature without.

If we do want to add it, here's the patch:

```
diff --git a/lib/core/base/cache.js b/lib/core/base/cache.js
index ec414ec6..d0773055 100644
--- a/lib/core/base/cache.js
+++ b/lib/core/base/cache.js
@@ -21,6 +21,10 @@ const cache = {
     const args = [...arguments];
     const hasDefaultValue = args.length === 2;
 
+    if (args.length > 2 || args.length === 0) {
+      throw new Error('incorrect number of arguments');
+    }
+
     if (
       hasDefaultValue &&
       !(key in _cache) &&
diff --git a/test/core/base/cache.js b/test/core/base/cache.js
index cb0befe9..b9f51018 100644
--- a/test/core/base/cache.js
+++ b/test/core/base/cache.js
@@ -96,4 +96,18 @@ describe('axe._cache', function () {
   it('should return undefined for key lookup miss when no default value is provided', function () {
     assert.strictEqual(typeof axe._cache.get('foo'), 'undefined');
   });
+
+  it('should throw if there are too many arguments', function () {
+    function fn() {
+      axe._cache.get('foo', 'bar', 'baz');
+    }
+    assert.throws(fn);
+  });
+
+  it('should throw if there are no arguments', function () {
+    function fn() {
+      axe._cache.get();
+    }
+    assert.throws(fn);
+  });
 });
```

**Review by WilcoFiers — CHANGES_REQUESTED** [review]
I'd like to see this get used as well. No point adding this if we're not going to use it.

File: lib/core/base/cache.js [hunk 1]
```diff
@@ -13,9 +13,24 @@ const cache = {
...
    * Retrieve an item from the cache.
    * @param {String} key - Name of the key the value was stored as.
+   * @param {*} defaultValue - Default value to set if there is a cache miss. Functions are evaluated before caching. To override a value already saved, use `set()`.
    * @returns {*} The item stored
    */
-  get(key) {
+  get(key, defaultValue) {
+    // undefined is a valid cache value, so figure this out from args instead of a simple truthy check
+    const args = [...arguments];
+    const hasDefaultValue = args.length === 2;
```
Comment (WilcoFiers) [inline lib/core/base/cache.js:22]: This seems simpler:

```suggestion
  get(key, ...args) {
    const defaultValue = args[0];
    const hasDefaultValue = args.length === 1;
```

**Comment by straker** [comment]
I would argue that it is not valid to set the cache with a value of `undefined`. `undefined` is the only way we know if the cache was not set before, so it should not be a valid default argument value either. That is, passing `undefined` would do nothing.

**Comment by dbowling** [comment]
Although passing `undefined` and passing nothing are equivalent when reading the value directly, they are not equivalent in this PR because of the arg spreading, which can determine the difference between the caller's usage. In one case it writes the key with a value of `undefined`, and in the other (no second arg passed) it would not set the key at all. 

The `set()` function does not distinguish though. It writes to the cache with no checks.

I would propose that if we consider `undefined` to not be a valid cache value, we codify that decision in both the `get()` and `set()` logic and tests as the current behavior would then be a bug. Probably via a separate issue/PR for the fix, and putting this on-hold until it is merged into develop.

**Comment by straker** [comment]
I rescind my comment on `undefined being a valid value.

File: lib/core/base/cache.js [hunk 2]
```diff
@@ -13,9 +13,24 @@ const cache = {
...
+
+    if (
+      hasDefaultValue &&
+      !(key in _cache) &&
+      typeof defaultValue === 'function'
+    ) {
+      this.set(key, defaultValue.call());
```
Comment (straker) [inline lib/core/base/cache.js:29]:
```suggestion
      this.set(key, defaultValue());
```

File: lib/core/base/cache.js [hunk 3]
```diff
@@ -13,9 +13,24 @@ const cache = {
...
+    } else if (hasDefaultValue && !(key in _cache)) {
+      this.set(key, defaultValue);
+    }
```
Comment (straker) [inline lib/core/base/cache.js:32]:
```suggestion
    if (hasDefaultValue && !(key in _cache)) {
      this.set(key, typeof defaultValue === 'function' ? defaultValue() : defaultValue);
    }
```

**Review by WilcoFiers — CHANGES_REQUESTED** [review]
This: https://github.com/dequelabs/axe-core/pull/3608#pullrequestreview-1074064357

**Comment by dbowling** [comment]
The following files in `lib` use `cache.set()`, but I elected not to include them in the refactor

- `lib/core/utils/get-flattened-tree.js` - only sets new values and does not use return
- `lib/core/public/run/globals-setup.js `- set needs to overwrite
- `lib/checks/generic/page-no-duplicate-evaluate.js` - refactor of this would likely be substantially harder to read

File: lib/core/base/cache.js [hunk 4]
```diff
@@ -13,9 +13,24 @@ const cache = {
   /**
    * Retrieve an item from the cache.
    * @param {String} key - Name of the key the value was stored as.
+   * @param {*} defaultValue - Default value to set if there is a cache miss. Functions are evaluated before caching. To override a value already saved, use `set()`.
    * @returns {*} The item stored
    */
-  get(key) {
+  get(key, ...args) {
+    const defaultValue = args[0];
+    const hasDefaultValue = args.length === 1;
```
Comment (stephenmathieson) [inline lib/core/base/cache.js:21]: Why do we need to rely on arity here? This smells like magic.

Comment (stephenmathieson) [reply lib/core/base/cache.js:21]: Also, what if someone accidentally passes 2 parameters to this function?

Comment (dbowling) [reply lib/core/base/cache.js:21]: Since `undefined` is treated as a value that could be cached, we can't just check for `undefined` of the second parameter. 

The best way I can think of doing it is to look at the length of arguments passed by the caller by spread or `arguments.length`.

The params have had a couple different ways of getting defined suggested. Personally, I'm not a huge fan of using the spread of args as the second parameter since only the first value would be used. I'll toss another one and see if that floats higher in the water for us:

```suggestion
    get(key, defaultValue) {
        const hasDefaultValue = arguments.length === 2;
```
IMO, a named parameter instead of `...args` addresses most of the magic smell. The other thing that seems more obvious to me is to make the second parameter an options object. `options = { default: null }`

If someone sends a second argument in accidentally, they would be setting the cache for that key if the key was not previously defined. It does not override an existing cache value though.

We could also leave `cache.get()` untouched and add Wilco's requested logic to a new method, something like `cache.getWithDefaultValue()`.

```suggestion
    getWithDefaultValue(key, defaultValue) {
        if( !get(key) ){
            set(key, defaultValue); 
            //plus function detection with .call()
        };
        return get(key);
```

Comment (stephenmathieson) [reply lib/core/base/cache.js:21]: Why would we need to cache `undefined`?

Comment (dbowling) [reply lib/core/base/cache.js:21]: My understanding is that Axe throws around `undefined` as a meaningful value by convention, and we may want to be caching that return. `cache.set()` will let you do so currently.

**Comment by dbowling** [comment]
I converted this PR back to draft while I work through the expanded scope of updating `lib` to use the default value poperty.

File: lib/core/base/cache.js (hunk 4 above)
Comment (stephenmathieson) [reply lib/core/base/cache.js:21]: axe-core uses `undefined` to denote an incomplete check. I can't think of any reason we'd cache that though.

Unless we have an actual use case for doing `cache.set('foo', undefined)`, I do not think we should support it.

**Comment by stephenmathieson** [comment]
What is the benefit of passing a default value? Why can't callers do:

```
thing = cache.get('thing') || defaultValue
```

File: lib/core/base/cache.js (hunk 4 above)
Comment (dbowling) [reply lib/core/base/cache.js:21]: Tagging @WilcoFiers and @straker

**Comment by dbowling** [comment]
Stephen, that gets back to what Wilco put in https://github.com/dequelabs/axe-core/issues/3276.

Your example `thing = cache.get('thing') || defaultValue` is close, but would need to also set the default value to the cache. In practice, that creates a pattern that us trying to be generalized.

Would this be more clear if we left `get()` as is, and added the new function call? This is a bit wordy, but gets the discussion started: `cache.getAndSetDefaultIfNotSet(key, default)`?

**Comment by stephenmathieson** [comment]
Looks like the issue is asking for:

```js
var thing = cache.pull('thing', () => {
  // This is only run if `thing` is not in the cache.
  return 7
})
```

This PR is not doing that. The issue also never makes any mention of `undefined`.

File: lib/core/base/cache.js (hunk 4 above)
Comment (straker) [reply lib/core/base/cache.js:21]: @stephenmathieson There are a few functions in axe-core where `undefined` is a valid cache value, such as [`is-modal-open](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/is-modal-open.js#L100-L101). The point of the cache is to denote an operation (potentially expensive) that we do not want to do again. What the operation returns is not important.

**Comment by dbowling** [comment]
Yes, that and a CI/local test discrepancy is why I took this back to draft status. If we want to change the intended behavior, thats okay too. Cache is a potential nightmare for bugs to hide, so I am all about being intentional. 

The undefined discussion started out of a review of finding appropriate test cases for the change. The code isn't currently explicitly defining how undefined should be treated and it is up to the caller. It appears there are mixed opinions on that, so it's good we are discussing as we evaluate how we want to use the cache.

File: lib/core/base/cache.js (hunk 4 above)
Comment (stephenmathieson) [reply lib/core/base/cache.js]: What is the difference between not setting anything in the cache and setting `undefined`? I don't understand why we'd ever need this.

Comment (straker) [reply lib/core/base/cache.js]: Setting anything in the cache means we've performed an operation and do not wish to do it again. If it helps, think of it as memoizing a function, but instead of caching the value for a set of parameters, we can cache the value for the entire run. It doesn't matter if the memoized function returned `undefined` (or anything really), the function will not be run again for the same set of parameters.

File: lib/checks/navigation/heading-order-evaluate.js [hunk 5]
```diff
@@ -3,6 +3,39 @@ import { querySelectorAllFilter, getAncestry } from '../../core/utils';
...
+  // find all headings, even ones that are outside the current
+  // context. also need to know where iframes are so we can insert
+  // the results of any in-context iframes into their proper order
+  // @see https://github.com/dequelabs/axe-core/issues/728
+
+  const headingOrder = cache.get('headingOrder', getHeadingOrder());
+
+  this.data({ headingOrder });
+
+  return true;
```
Comment (straker) [inline lib/checks/navigation/heading-order-evaluate.js:18]: I would say this function doesn't benefit from the cache get default value pattern. The main reason is that the setting of `data` should happen inside `getHeadingOrder` and not in the evaluate function (we only want the data on the first node, not every node), which means you need to pass `this` to `getHeadingOrder` or use an arrow function instead of a named function. That complicates things and only adds further nesting, so I think we should revert and leave this case as is.

File: lib/checks/navigation/region-evaluate.js [hunk 6]
```diff
@@ -81,38 +83,31 @@ function findRegionlessElms(virtualNode, options) {
...
-      return vNode;
-    })
-    // Remove duplicate containers
-    .filter((vNode, index, array) => {
-      return array.indexOf(vNode) === index;
-    });
-  cache.set('regionlessNodes', regionlessNodes);
+  const regionlessNodes = cache.get(
+    'regionlessNodes',
+    findRegionlessElms(axe._tree[0], options)
```
Comment (straker) [inline lib/checks/navigation/region-evaluate.js:92]: I think I would prefer this as it's own function for how complex it is.

```js
const regionlessNodes = cache.get(
  'regionlessNodes',
  () => getRegionlessNodes(options)
);

function getRegionlessNodes(options) {
  regionlessNodes = findRegionlessElms(axe._tree[0], options)
    // Find first parent marked as having region descendant (or body) and
    // return the node right before it as the "outer" element
    .map(vNode => {
      while (
        vNode.parent &&
        !vNode.parent._hasRegionDescendant &&
        vNode.parent.actualNode !== document.body
      ) {
        vNode = vNode.parent;
      }

      return vNode;
    })
    // Remove duplicate containers
    .filter((vNode, index, array) => {
      return array.indexOf(vNode) === index;
    });
}
```

File: lib/commons/dom/is-modal-open.js [hunk 7]
```diff
@@ -34,26 +37,30 @@ function isModalOpen(options) {
...
+function definiteModalsExists(roles) {
+  return Boolean(
+    querySelectorAllFilter(
+      // TODO: es-module-_tree
+      axe._tree[0],
+      roles,
+      vNode => isVisible(vNode.actualNode)
+    ).length
+  );
+}
```
Comment (straker) [inline lib/commons/dom/is-modal-open.js:56]: This is an antipattern to use the `Boolean` constructor. Instead return a comparison.

```suggestion
  return querySelectorAllFilter(
      // TODO: es-module-_tree
      axe._tree[0],
      roles,
      vNode => isVisible(vNode.actualNode)
    ).length > 0;
}
```

File: lib/commons/dom/is-modal-open.js [hunk 8]
```diff
@@ -34,26 +37,30 @@ function isModalOpen(options) {
...
-
-  const definiteModals = querySelectorAllFilter(
-    // TODO: es-module-_tree
-    axe._tree[0],
-    'dialog, [role=dialog], [aria-modal=true]',
-    vNode => isVisible(vNode.actualNode)
+  return cache.get(
+    'isModalOpen',
+    definiteModalsExists(definiteModalRoles) ||
+      probableModalExists(modalPercent)
```
Comment (straker) [inline lib/commons/dom/is-modal-open.js:43]: I feel that this change could lead to causing a subtle bug. If the order of the OR conditions were changed, the code would not work as expected since `probableModalExists` would return `undefined` and then still execute `definiteModalsExists`

File: lib/commons/dom/is-skip-link.js [hunk 9]
```diff
@@ -54,3 +35,22 @@ export default function isSkipLink(element) {
...
+  } else {
+    firstPageLink = querySelectorAll(
+      axe._tree,
+      'a[href]:not([href^="javascript:"])'
+    ).find(link => !isCurrentPageLink(link.actualNode));
+  }
+
+  // null will signify no first page link
+  cache.set('firstPageLink', firstPageLink || null);
+  return firstPageLink;
```
Comment (straker) [inline lib/commons/dom/is-skip-link.js:55]:
```suggestion
  return firstPageLink || null;
```

File: lib/core/base/cache.js [hunk 10]
```diff
@@ -13,10 +13,21 @@ const cache = {
   /**
    * Retrieve an item from the cache.
    * @param {String} key - Name of the key the value was stored as.
+   * @param {*} defaultValue - Default value to set if there is a cache miss. Functions are evaluated before caching. To override a value already saved, use `set()`.
```
Comment (straker) [inline lib/core/base/cache.js:16]:
```suggestion
   * @param {*} [defaultValue] - Default value to set if there is a cache miss. Functions are evaluated before caching. To override a value already saved, use `set()`.
```

File: lib/checks/navigation/heading-order-evaluate.js [hunk 11]
```diff
@@ -3,6 +3,39 @@ import { querySelectorAllFilter, getAncestry } from '../../core/utils';
 import { isVisible } from '../../commons/dom';
 import { getRole } from '../../commons/aria';
 
+export default headingOrderEvaluate;
+
+function headingOrderEvaluate() {
```
Comment (straker) [inline lib/checks/navigation/heading-order-evaluate.js:8]:
```suggestion
export default function headingOrderEvaluate() {
```

File: lib/core/utils/filter-html-attrs.js [hunk 12]
```diff
@@ -73,3 +62,18 @@ function filterHtmlAttrs(element, filterAttrs) {
 }
 
 export default filterHtmlAttrs;
```
Comment (straker) [inline lib/core/utils/filter-html-attrs.js:64]: The default export function itself should be moved to the top with an inlined `default export`.

File: test/core/base/cache.js [hunk 13]
```diff
@@ -1,14 +1,18 @@
-describe('axe._cache', function() {
+describe('axe._cache', function () {
   'use strict';
 
-  it('should set items from the cache without error', function() {
+  beforeEach(function () {
+    axe._cache.clear();
+  });
+
```
Comment (straker) [inline test/core/base/cache.js:7]: This is handled by a [global `afterEach` in testutils](https://github.com/dequelabs/axe-core/blob/develop/test/testutils.js#L565)

```suggestion
```

**Review by straker — CHANGES_REQUESTED** [review]

File: lib/core/base/cache.js (hunk 4 above)
Comment (stephenmathieson) [reply lib/core/base/cache.js]: Perhaps I don't understand the code, but it looks like the function will run again if `undefined` is cached based on these lines: https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/is-modal-open.js#L37-L39

Comment (straker) [reply lib/core/base/cache.js]: Thanks for pointing that out, looks like a bug in the `is-modal-open` function.

Comment (stephenmathieson) [reply lib/core/base/cache.js]: I still don't believe there's a reason to cache `undefined`.

File: lib/commons/dom/is-skip-link.js [hunk 14]
```diff
@@ -54,3 +35,21 @@ export default function isSkipLink(element) {
     element.DOCUMENT_POSITION_FOLLOWING
   );
 }
+
+function generateFirstPageLink(firstPageLink) {
```
Comment (WilcoFiers) [inline lib/commons/dom/is-skip-link.js:39]:
```suggestion
function generateFirstPageLink() {
    let firstPageLink;
```

File: lib/core/base/cache.js [hunk 15]
```diff
@@ -13,10 +13,21 @@ const cache = {
...
+  get(key, defaultValue) {
+    if (key in _cache) {
+      return _cache[key];
+    }
+
+    if (arguments.length === 2) {
+      this.set(
+        key,
+        typeof defaultValue === 'function' ? defaultValue() : defaultValue
+      );
```
Comment (WilcoFiers) [inline lib/core/base/cache.js:28]: I think we may need to throw if `creator` returns undefined. Because if it does we're just going to end up rerunning the creator method until it doesn't. That kind of defeats the purpose of the cache. Seeing how easy it is to get this wrong, I think it'd be safer not to allow undefined.

I spotted we had a bug because we're allowing this in isModalOpen (see my comments there).

File: lib/commons/dom/is-modal-open.js [hunk 16]
```diff
@@ -34,26 +40,32 @@ function isModalOpen(options) {
...
   // just our best guess of being one based on convention.
 
-  if (cache.get('isModalOpen')) {
-    return cache.get('isModalOpen');
+  if (definitelyModalExists()) {
+    return true;
+  } else if (couldBeModalExists(modalPercent)) {
+    return true;
   }
+  return undefined;
```
Comment (WilcoFiers) [inline lib/commons/dom/is-modal-open.js:48]: If we pass undefined into the cache, it's not going to know this has been set. `detectModal` is going to run over and over again.

```suggestion
  return false;
```

**Review by WilcoFiers — CHANGES_REQUESTED** [review]

File: lib/core/base/cache.js [hunk 17]
```diff
@@ -13,10 +13,21 @@ const cache = {
   /**
    * Retrieve an item from the cache.
    * @param {String} key - Name of the key the value was stored as.
+   * @param {*} [defaultValue] - Default value to set if there is a cache miss. Functions are evaluated before caching. To override a value already saved, use `set()`.
    * @returns {*} The item stored
    */
-  get(key) {
-    return _cache[key];
+  get(key, defaultValue) {
```
[the remark this answers is not in the document]
Comment (dbowling) [reply lib/core/base/cache.js:21]: > I'm coming back to the idea of allowing a default that isn't a function. Just in this PR, there are a dozen places where it wasn't used correctly. I can't think of a single use case where this shouldn't be a function. I think we're better off enforcing a function. If we have any case where this shouldn't be a function, we can always use `.set()`.

I'm concerned that allowing `set()` to bypass the validation rules might led to unintentional bugs. I'm going to take a little inspiration from the React team and try to deliver your goal, but with protections by default and including an escape hatch with something like `dangerouslySet()`.

File: test/core/base/cache.js [hunk 18]
```diff
@@ -1,32 +1,160 @@
...
+          axe._cache.get('boolean', function () {
+            return true;
+          })
+        );
+        assert.isNull(
+          axe._cache.get('null', function () {
+            return null;
+          })
+        );
+        assert.deepEqual(
```
Comment (WilcoFiers) [inline test/core/base/cache.js:85]: We should check equality, not just deepEqual. We want it to be the same object that's returned.

File: lib/core/base/cache.js [hunk 19]
```diff
@@ -7,16 +7,27 @@ const cache = {
...
    * @returns {*} The item stored
    */
-  get(key) {
-    return _cache[key];
+  get(key, creator) {
+    if (key in _cache) {
+      return _cache[key];
+    }
+
+    if (creator) {
```
Comment (WilcoFiers) [inline lib/core/base/cache.js:26]:
```suggestion
    if (typeof creator !== 'undefined') {
```

File: test/core/base/cache.js [hunk 20]
```diff
@@ -1,32 +1,160 @@
...
+        assert.isBoolean(
+          axe._cache.get('Foo', function () {
+            return 'should not be set to a string because Foo is in the cache';
+          })
+        );
+      });
+
+      it('should not override a value from `set()`', function () {
+        axe._cache.set('foo', 'bar');
+        axe._cache.get('foo', 'baz');
```
Comment (WilcoFiers) [inline test/core/base/cache.js:138]: It's weird for this not to throw. I think we should always check the creator type, regardless of if there's a value already or not. It's odd for this function to accept an invalid argument if a value was previously set.

File: lib/commons/dom/is-modal-open.js [hunk 21]
```diff
@@ -14,6 +16,14 @@ function isModalOpen(options) {
   options = options || {};
   const modalPercent = options.modalPercent || 0.75;
 
+  try {
```
Comment (WilcoFiers) [inline lib/commons/dom/is-modal-open.js:19]: What's this for?

File: lib/commons/dom/is-skip-link.js [hunk 22]
```diff
@@ -54,3 +35,22 @@ export default function isSkipLink(element) {
     element.DOCUMENT_POSITION_FOLLOWING
   );
 }
+
+function generateFirstPageLink() {
+    let firstPageLink;
```
Comment (WilcoFiers) [inline lib/commons/dom/is-skip-link.js:40]:
```suggestion
  let firstPageLink;
```

**Review by WilcoFiers — DISMISSED** [review]

File: lib/commons/dom/is-modal-open.js (hunk 21 above)
Comment (dbowling) [reply lib/commons/dom/is-modal-open.js:19]: The creator `detectModal` returns undefined on line `52`, which causes `cache.get` to throw. That looks to be entirely normal behavior with this, so the error is being caught.

I can toss in a comment, check for that specific error type, or both. It doesn't look like we would want to change the return behavior.

File: lib/commons/dom/is-skip-link.js (hunk 22 above)
Comment (dbowling) [reply lib/commons/dom/is-skip-link.js:40]: Strange, I guess my prettier didn't work on this commit.

File: test/core/base/cache.js (hunk 20 above)
Comment (dbowling) [reply test/core/base/cache.js:141]: I'll update to do that check, though we won't be able to do 100% of the validation every time because that would require executing the function to determine the return type (since `undefined` is not allowed.)

Comment (dbowling) [reply test/core/base/cache.js]: I was mistaken about your intent on undefined so ignore that comment. This is now resolved.

File: lib/commons/dom/is-modal-open.js (hunk 21 above)
Comment (dbowling) [reply lib/commons/dom/is-modal-open.js]: As above, I misread the date that your comment on undefined was. This try/catch comes out when considering that so I am resolving.

File: test/core/base/cache.js [hunk 23]
```diff
@@ -1,32 +1,173 @@
...
+      }
+      assert.doesNotThrow(fn);
+    });
+
+    // undefined is a valid result type of a potentially expensive operation in axe-core
+    it('should set `undefined` without error', function () {
+      axe._cache.set('foo', undefined);
+
+      function fn() {
+        assert.isUndefined(axe._cache.get('foo'));
```
Comment (straker) [inline test/core/base/cache.js:17]: You're testing the `get` function in the `set` describe block. We should make sure that setting `undefined` does not throw, not that `get` does not throw

```suggestion
      function fn() {
        axe._cache.set('foo', undefined);
```

File: test/core/base/cache.js [hunk 24]
```diff
@@ -1,32 +1,173 @@
...
+            return obj;
+          }),
+          obj
+        );
+
+        assert.sameOrderedMembers(
+          axe._cache.get('array', function () {
+            return [1, 2, 3];
+          }),
+          [1, 2, 3]
```
Comment (straker) [inline test/core/base/cache.js:100]: You'll want to check that it returns the same array like you did for the obj

```suggestion
        var array = [1, 2, 3];
        assert.equal(
          axe._cache.get('array', function () {
            return array;
          }),
          array
```

**Review by straker — DISMISSED** [review]

File: test/core/base/cache.js [hunk 25]
```diff
@@ -1,32 +1,180 @@
...
-  });
+  describe('set', function () {
+    it('should set items without error', function () {
+      function fn() {
+        axe._cache.set('foo', 'bar');
+      }
+      assert.doesNotThrow(fn);
+    });
+
+    // undefined is a valid result type of a potentially expensive operation in axe-core
```
Comment (stephenmathieson) [inline test/core/base/cache.js:12]: You all have convinced me that you believe this is true. You have yet to provide a valid use case for caching nothing though. Why is this necessary?

**Review by stephenmathieson — CHANGES_REQUESTED** [review]

File: test/core/base/cache.js (hunk 25 above)
Comment (straker) [reply test/core/base/cache.js:12]: Both @WilcoFiers and I have agreed that we want this functionality

Comment (stephenmathieson) [reply test/core/base/cache.js:12]: Why? I don't understand why an explanation is too much to ask.

File: lib/core/base/cache.js [hunk 26]
```diff
@@ -7,16 +7,29 @@ const cache = {
...
-    return _cache[key];
+  get(key, creator) {
+    validateCreator(creator);
+
+    if (key in _cache) {
+      return _cache[key];
+    }
+
+    if (typeof creator === 'function') {
+      const value = creator();
```
Comment (WilcoFiers) [inline lib/core/base/cache.js:29]: Did we not say we were going to do this?

```suggestion
      const value = creator();
      assert(value !== 'undefined', 'Cache creator function should not return undefined');
```

File: lib/core/base/cache.js [hunk 27]
```diff
@@ -27,4 +40,35 @@ const cache = {
...
+ * @returns {Void} Void if the key is valid.
+ * @throws {TypeError} If the key is not a string.
+ */
+function validateKey(key) {
+  if (typeof key !== 'string') {
+    throw new TypeError('key must be a string, ' + typeof key + ' given');
+  }
+  if (key === '') {
+    throw new TypeError('key must not be empty');
+  }
```
Comment (WilcoFiers) [inline lib/core/base/cache.js:55]: Did you know axe-core has an `assert`? Might be good to use:

```suggestion
  assert(typeof key !== 'string', 'key must be a string, ' + typeof key + ' given');
  assert(key === '', 'key must not be empty');
```

**Review by WilcoFiers — DISMISSED** [review]

File: lib/core/base/cache.js (hunk 26 above)
Comment (stephenmathieson) [reply lib/core/base/cache.js:29]: Without the extra quotes:

```suggestion
      const value = creator();
      assert(value !== undefined, 'Cache creator function should not return undefined')
```

File: test/core/base/cache.js [hunk 28]
```diff
@@ -9,35 +9,34 @@ describe('axe._cache', function () {
...
-    // undefined is a valid result type of a potentially expensive operation in axe-core
     it('should set `undefined` without error', function () {
       function fn() {
         axe._cache.set('foo', undefined);
       }
       assert.doesNotThrow(fn);
     });
 
-    it('should throw TypeError for non-string keys', function () {
+    it('should throw for non-string keys', function () {
```
Comment (stephenmathieson) [inline test/core/base/cache.js:19]: Probably not a big deal, but IMO this should still be a `TypeError` (the error is due to the wrong type being used).

Comment (dbowling) [reply test/core/base/cache.js:19]: Agreed, I would rather throw a more specific error. I changed this because I was asked to use the internal assert, which only throws errors.

@straker and @WilcoFiers  How would you feel about me adding an error type param to the axe assert helper so we can continue to assert the more specific type?

**Comment by dbowling** [comment]
@straker or @WilcoFiers  - how would you like me to handle the current failed tests for places where the [creator is returning undefined and throwing an error](https://circleci.com/gh/dequelabs/axe-core/48476?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)?

If we want to continue to throw, I'll need to touch these functions and check before sending, put in try/catch, or add an options object that permits cache to quietly ignore (e.g. not setting a value and not throwing) when it encounters this case.

Happy to do whatever you want, I'm mindful of scope bloat on this PR though. If that scope increase concerns you, we can still remove the throw (or make it a non-default setting) as that is new behavior for this PR.

File: lib/core/base/cache.js (hunk 26 above)
Comment (stephenmathieson) [reply lib/core/base/cache.js:31]: Asked on Slack (but didn't get a response), so I'll do it again here...

If we are not allowing:

```js
cache.get('foo', function(){ return undefined })
```

Then why do we allow the below?

```js
cache.set('foo', undefined)
```

File: test/core/base/cache.js (hunk 28 above)
Comment (dbowling) [reply test/core/base/cache.js:19]: I spoke with Wilco and Steve earlier today and discussed adding an error type to the internal assert helper to improve test quality. The rest of axe-core doesn't differentiate between error types, so they would rather not change the pattern for just the cache. I prefer the more specific error types, but it is probably a solution in search of a problem here.

**Comment by dbowling** [comment]
The [reply to my last comment can be found here](https://github.com/dequelabs/axe-core/pull/3608#discussion_r966448810).

Since the `assert` is throwing for `is-modal-open`, we also decided that a better solution than treating that as a bug is to memoize.

File: lib/core/base/cache.js (hunk 26 above)
Comment (dbowling) [reply lib/core/base/cache.js:31]: The assert helper has been integrated. Stephen, I think Steven got to chat with you on Slack last week. Thanks for the feedback, but it looks like we're going anther direction on caching undefined so I'm going to resolve this conversation to get the PR ready for re-review with current scope.

Comment (stephenmathieson) [reply lib/core/base/cache.js:31]: The question above was never answered (neither on Slack nor here).

File: test/core/base/cache.js (hunk 25 above)
Comment (dbowling) [reply test/core/base/cache.js]: My understanding is that the two of you spoke offline on this question. I'm resolving this thread so we can finalize review.

Comment (stephenmathieson) [reply test/core/base/cache.js]: This has still not been resolved. We did speak on Slack, but the question posted at https://github.com/dequelabs/axe-core/pull/3608#discussion_r965041066 has never been answered.

File: lib/commons/dom/is-modal-open.js [hunk 29]
```diff
@@ -10,7 +10,7 @@ import { querySelectorAllFilter } from '../../core/utils';
  * @instance
  * @return {Boolean|undefined} True if we know (or our best guess) that a modal is open, undefined if we can't tell (doesn't mean there isn't one open)
  */
-function isModalOpen(options) {
+export default memoize(function isModalOpen(options) {
```
Comment (WilcoFiers) [inline lib/commons/dom/is-modal-open.js:13]: The options object is going to prevent memoize from working. This PR isn't about memoization anyway, so I don't think this belongs here. If we can't use the creator method for isModalOpen, best to just leave it as is.

File: lib/core/utils/filter-html-attrs.js [hunk 30]
```diff
@@ -2,25 +2,6 @@ import cache from '../base/cache';
 import getNodeAttributes from './get-node-attributes';
 import matchesSelector from './element-matches';
 
-/**
```
Comment (WilcoFiers) [inline lib/core/utils/filter-html-attrs.js:5]: Why was this file changed?

**Review by WilcoFiers — CHANGES_REQUESTED** [review]

File: lib/core/utils/filter-html-attrs.js (hunk 30 above)
Comment (dbowling) [reply lib/core/utils/filter-html-attrs.js:5]: I was confused at first too, but there's nothing special about the changes to this file. It refactors the same thing the rest of them do, removing this pattern:

```JavaScript
if (cache.get(outerHTML)) {
    node = cache.get(outerHTML);
  }
```

It looks strange because the first function wasn't the default export, so I had switched their order to comply with the new code standards.

File: test/core/base/cache.js [hunk 31]
```diff
@@ -1,32 +1,169 @@
...
+          axe._cache.get('foo', function () {
+            return undefined;
+          });
+        }
+        assert.throws(fn);
+      });
+
+      it('should not evaluate creator if key already exists', function () {
+        var spy = sinon.spy();
+        axe._cache.set('foo', spy);
```
Comment (WilcoFiers) [inline test/core/base/cache.js:120]: Passing the spy in here is confusing.

```suggestion
        axe._cache.set('foo', 'bar');
```

File: test/core/base/cache.js [hunk 32]
```diff
@@ -1,32 +1,169 @@
...
+        axe._cache.get('foo', spy);
+        assert.isFalse(spy.calledOnce);
```
Comment (WilcoFiers) [inline test/core/base/cache.js:122]: This makes it look as though this could be called more than once. It shouldn't have been called at all.

**Review by WilcoFiers — CHANGES_REQUESTED** [review]
Couple minor points

File: test/core/base/cache.js (hunk 31 above)
Comment (dbowling) [reply test/core/base/cache.js]: Makes sense, thanks for the feedback.

File: test/core/base/cache.js (hunk 32 above)
Comment (dbowling) [reply test/core/base/cache.js]: Good point, this has been clarified.

**Review by WilcoFiers — APPROVED** [review]
